plugins/inputs/mavlink/mavlink_utils.go

```diff
@@ -0,0 +1,136 @@
+package mavlink
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/chrisdalke/gomavlib/v3"
+)
+
+// Convert a string from CamelCase to snake_case
+func ConvertToSnakeCase(input string) string {
+	re := regexp.MustCompile(`([a-z0-9])([A-Z])`)
+	snake := re.ReplaceAllString(input, `${1}_${2}`)
+	snake = strings.ToLower(snake)
+	return snake
+}
+
+// Check if a string is in a slice
+func Contains(slice []string, str string) bool {
+	for _, item := range slice {
+		if item == str {
+			return true
+		}
+	}
+	return false
+}
+
+// Convert a Mavlink event into a struct containing Metric data.
+func MavlinkEventFrameToMetric(frm *gomavlib.EventFrame) metricFrameData {
+	out := metricFrameData{}
+	out.tags = make(map[string]string)
+	out.fields = make(map[string]any)
+	out.tags["sys_id"] = strconv.FormatUint(uint64(frm.SystemID()), 10)
+
+	m := frm.Message()
+	t := reflect.TypeOf(m)
+	v := reflect.ValueOf(m)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		v = v.Elem()
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		value := v.Field(i)
+		out.fields[ConvertToSnakeCase(field.Name)] = value.Interface()
+	}
+
+	out.name = ConvertToSnakeCase(t.Name())
+	out.name = strings.TrimPrefix(out.name, "message_")
+	return out
+}
+
+// Parse the FcuURL config to setup a mavlib endpoint config
+func ParseMavlinkEndpointConfig(s *Mavlink) ([]gomavlib.EndpointConf, error) {
+	if strings.HasPrefix(s.FcuURL, "serial://") {
+		tmpStr := strings.TrimPrefix(s.FcuURL, "serial://")
+		tmpStrParts := strings.Split(tmpStr, ":")
+		deviceName := tmpStrParts[0]
```

Comment:
Go builtin `net/url` has useful `Parse` method to facilitate this easier..

Reply:
Updated to remove `strings.HasPrefix(s.FcuURL, "serial://")` and use `url.Parse` instead. There's still a little string find-and-replace here, because the serial port paths have forward slashes in them, which seems to mess with the URL parsing. I think the new code is simple enough.